spring-integration-mongodb/src/main/java/org/springframework/integration/mongodb/outbound/ReactiveMongoDbStoringMessageHandler.java

```diff
@@ -0,0 +1,131 @@
+/*
+ * Copyright 2019 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.springframework.integration.mongodb.outbound;
+
+import org.springframework.data.mongodb.MongoDbFactory;
+import org.springframework.data.mongodb.ReactiveMongoDatabaseFactory;
+import org.springframework.data.mongodb.core.MongoOperations;
+import org.springframework.data.mongodb.core.ReactiveMongoOperations;
+import org.springframework.data.mongodb.core.ReactiveMongoTemplate;
+import org.springframework.data.mongodb.core.convert.MongoConverter;
+import org.springframework.expression.Expression;
+import org.springframework.expression.common.LiteralExpression;
+import org.springframework.expression.spel.support.StandardEvaluationContext;
+import org.springframework.integration.expression.ExpressionUtils;
+import org.springframework.integration.handler.AbstractReactiveMessageHandler;
+import org.springframework.messaging.Message;
+import org.springframework.util.Assert;
+
+import reactor.core.publisher.Mono;
+
+/**
+ * Implementation of {@link org.springframework.messaging.ReactiveMessageHandler} which writes
+ * Message payload into a MongoDb collection, using reactive MongoDb support, The
+ * collection is identified by evaluation of the {@link #collectionNameExpression}.
+ *
+ * @author David Turanski
+ *
+ * @since 5.3
+ */
+public class ReactiveMongoDbStoringMessageHandler extends AbstractReactiveMessageHandler {
+
+	private volatile ReactiveMongoOperations mongoTemplate;
+
+	private volatile ReactiveMongoDatabaseFactory mongoDbFactory;
+
+	private volatile MongoConverter mongoConverter;
+
+	private volatile StandardEvaluationContext evaluationContext;
+
+	private volatile Expression collectionNameExpression = new LiteralExpression("data");
+
+	private boolean initialized = false;
```

Comment:
No, you didn't get me.
I asked exactly about opposite. All other properties must go without a `volatile` and only this one has to have it, since only this one is definitely accessed from different threads. All others are populated from a single thread and never (roughly) changed.

Reply:
`evaluationContext` and `collectionNameExpression` are also accessed from different threads

Reply:
That's true, but the point is how they are changes. It is definitely a nature of SI components that they are accessed from different threads. Only a difference how they are changed at runtime.
I mean independently of access we set them only once - during application start up.

Reply:
I think SI is the only Spring project that prefers to use volatile. I'm not sure that we ever encountered an actual race condition where it made a difference but I always put it in SI classes for legacy reasons. You're the lead now so you can change it.